src/contract.rs

```diff
@@ -125,27 +153,129 @@ pub fn handle<S: Storage, A: Api, Q: Querier>(
     let contract_status = ReadonlyConfig::from_storage(&deps.storage).contract_status();
 
     match contract_status {
-        ContractStatusLevel::StopAll | ContractStatusLevel::StopAllButRedeems => {
-            let response = match msg {
-                HandleMsg::SetContractStatus { level, .. } => set_contract_status(deps, env, level),
-                HandleMsg::Redeem { amount, .. }
-                    if contract_status == ContractStatusLevel::StopAllButRedeems =>
-                {
-                    try_redeem(deps, env, amount)
+        ContractStatusLevel::NormalRun => {} // If it's a normal run just continue
+        _ => {
+            let mut notAuthorized = false;
+            let status_code = status_level_to_u8(contract_status);
+
+            match msg.clone() {
+                // This is always allowed
+                HandleMsg::SetContractStatus {..} => {},
+                HandleMsg::UpdateStakeConfig {..} => {},
+
+                // If receive check that msg is not bonding or reward
+                HandleMsg::Receive {msg, ..} => {
+                    let receive_type: ReceiveType;
+                    if let Some(msg) = msg {
+                        receive_type = from_binary(&msg)?;
+                    }
+                    else {
+                        return Err(StdError::generic_err("No receive type supplied in message"))
+                    }
```

Comment:
no, a user will not be triggering receive on the Snip20 unless theyre doing one of the three defined actions. Remember that receive is only called when the receiving contract has registered receive (AKA user sending to the staking contract)

Reply:
no, a user will not be triggering receive on the Snip20 unless theyre doing one of the three defined actions. Remember that receive is only called when the receiving contract has registered receive (AKA user sending to the staking contract)

Reply:
How do you unbond your SHD? I understand its a snip-20, you had told me you were putting the contracts together into one, which I think they should be separate but if thats the case we don't need to make any changes to the snip-20. Is this the staking contract or just a snip-20 with a separate contract for staking management?

Reply:
This is the snip20 with the staking, unbonding is done by calling the unbond handle msg, since the contract is the snip20 then you dont need to send them, it just edits the internal state, same as burn for example

Reply:
IMO these should be separate contracts, and the managing contract should receive the share token and burn them as the underlying collateral is unbonded. It would require minimal changes to the snip-20, only things like denying transfer. But if this is the design you want i'll approve.